v2/internal/identity/credential_provider.go

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package identity
+
+import (
+	"context"
+	"reflect"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	"github.com/Azure/azure-service-operator/v2/internal/config"
+	"github.com/Azure/azure-service-operator/v2/internal/reconcilers"
+	"github.com/Azure/azure-service-operator/v2/internal/util/kubeclient"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/core"
+)
+
+const (
+	// #nosec
+	globalCredentialSecretName = "aso-controller-settings"
+	// #nosec
+	NamespacedSecretName = "aso-credential"
+)
+
+// Credential describes a credential used to connect to Azure
+type Credential struct {
+	TokenCredential azcore.TokenCredential
+	CredentialFrom  types.NamespacedName
+	SubscriptionID  string
```

Comment:
Should these be publicly visible (and thus mutable), or should you create some getter methods to safely expose them?

Reply:
Moved to methods